File: 17.电话号码的字母组合.cpp
/*
 * @lc app=leetcode.cn id=17 lang=cpp
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(digits.length() == 0) return res;
        return dfs(digits, "", 0);
    }
    vector<string> res;
    map<char,string> key= {
        {'2', "abc"}, 
        {'3', "def"},
        {'4', "ghi"},
        {'5', "jkl"},
        {'6', "mno"},
        {'7', "pqrs"},
        {'8', "tuv"},
        {'9', "wxyz"},
    };
    vector<string> dfs(string digits, string temp, int len){
        if(len == digits.size()) res.push_back(temp);
        else{
            for(int i=0;i<key[digits[len]].length();i++)
            {
                temp += key[digits[len]][i];
                dfs(digits, temp, len+1);
                temp.pop_back();
            }
        }
        return res;
    }
};
// @lc code=end

File: 17.电话号码的字母组合.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=17 lang=java
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
    public List<String> letterCombinations(String digits) {
        key = new HashMap<>();
        key.put('2', "abc");
        key.put('3', "def");
        key.put('4', "ghi");
        key.put('5', "jkl");
        key.put('6', "mno");
        key.put('7', "pqrs");
        key.put('8', "tuv");
        key.put('9', "wxyz");
        if(digits.length() == 0) return res;
        return dfs(digits, new StringBuilder(""), 0);
    }
    List<String> res = new ArrayList<>();
    private Map<Character, String> key;

    List<String> dfs(String digits, StringBuilder temp, int len){
        if(len == digits.length()) res.add(temp.toString());
        else{
            for(int i=0;i<key.get(digits.charAt(len)).length();i++)
            {
                temp.append(key.get(digits.charAt(len)).charAt(i));
                dfs(digits, temp, len+1);
                temp.deleteCharAt(temp.length()-1);
            }
        }
        return res;
    }
}
// @lc code=end

